sound/soc/sof/ipc4-pcm.c

```diff
@@ -940,20 +959,33 @@ static snd_pcm_sframes_t sof_ipc4_pcm_delay(struct snd_soc_component *component,
 	 * Firmware calculates correct stream_start_offset for all cases including above two.
 	 * Driver subtracts stream_start_offset from dai dma position to get accurate one
 	 */
-	tmp_ptr -= time_info->stream_start_offset;
 
-	/* Calculate the delay taking into account that both pointer can wrap */
-	div64_u64_rem(tmp_ptr, substream->runtime->boundary, &tmp_ptr);
+	/*
+	 * If the dai position has passed the start offset it means that the
+	 * audio left the DSP and we need to normalize the dai position
+	 * Otherwise the first audio frames is still within the DSP, we should
+	 * adjust the host side position to accurately report the initial delay
+	 */
+	if (tmp_ptr >= time_info->stream_start_offset) {
```

Comment:
isn't stream_start_offset only valid for playback?

Reply:
Yes, it looks like it is 0 for capture. In that case we report the difference between the host and dai counters w/o compensation.